pubtools/exodus/_hooks/pulp.py

```diff
@@ -0,0 +1,210 @@
+
+        for user_type, ident in (
+            ("client", "serviceAccountId"),
+            ("user", "internalUsername"),
+        ):
+            typed_ctx = context[user_type]
+            if typed_ctx["authenticated"]:
+                LOG.debug(
+                    "Authenticated with exodus-gw at %s as %s %s (roles: %s)",
+                    self._gw_url,
+                    user_type,
+                    typed_ctx[ident],
+                    [str(role) for role in typed_ctx["roles"]],
+                )
+                break
+        else:
+            LOG.debug("Not authenticated with exodus-gw.")
+
+    def _poll_commit_completion(self, commit, tries=4):
+        """Issues request(s) to exodus-gw to get the task state, returning once
+        it either is FAILED or COMPLETED."""
+        task_url = urljoin(self._gw_url, commit["links"]["self"])
+
+        for i in range(tries):
+            resp = self._do_request(method="GET", url=task_url)
+            task = resp.json()
+
+            if task["state"] not in ("FAILED", "COMPLETE"):
+                if i < tries - 1:
+                    wait_time = (i + 1) * 30 if wait_time < 120 else wait_time
+                    LOG.info(
+                        "Commit %s to exodus-gw at %s in progress. "
+                        "Will repoll in %d seconds [try %s/%s]",
+                        task["id"],
+                        self._gw_url,
+                        wait_time,
+                        i + 1,
+                        tries,
+                    )
+                    time.sleep(wait_time)
+                    continue
+
+                raise RuntimeError(
+                    "Commit %s to exodus-gw at %s timed out"
+                    % (task["id"], self._gw_url)
+                )
+            if task["state"] == "FAILED":
+                raise RuntimeError(
+                    "Commit %s to exodus-gw at %s failed"
+                    % (task["id"], self._gw_url)
+                )
+            if i != 0:
+                LOG.info(
+                    "Exodus config deployment succeeded [try: %s/%s]",
+                    i + 1,
+                    tries,
+                )
+            return task
+
+    @hookimpl
+    def pulp_repository_pre_publish(self, repository, options):
+        """Invoked before a Pulp repository is published.
+
+        This implementation adds to each config the --exodus-publish argument,
+        attaching the repository to an Exodus gateway publish.
+
+        Args:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of Distributor-config pairs intended for publishing.
+
+        Returns:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of potentially adjusted Distributor-config pairs to publish.
+        """
+
+        if not self._exodus_enabled():
+            return None
+
+        self._check_cert()
```

Comment:
I don't think you want to do this every time the hook is invoked.  If there are 100 repos being published, we don't need to check the cert 100 times.  Could we do this just within the "if not self._publish" block?